[SPARK-58519][SQL] Document UPDATE and DELETE FROM statements (#57725)
apache/spark: docs/sql-ref-syntax-dml-update.md

  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
---

### Description

The `UPDATE` statement changes the values of columns in rows that satisfy an optional condition.
When no condition is specified, every row is updated.

`UPDATE` is supported on tables backed by
[Data Source V2](sql-v2-data-sources.html#row-level-dml) connectors that support row-level
operations.

### Syntax

```sql
UPDATE table_identifier [ [ AS ] table_alias ]
    [ WITH ( option_key = option_value [ , ... ] ) ]
    SET column = expression [ , ... ]
    [ WHERE boolean_expression ]
```

### Parameters

* **table_identifier**

    Specifies the table to update, which may be optionally qualified with a database name.

    **Syntax:** `[ database_name. ] table_name`

* **table_alias**

    Specifies an optional alias for the target table. The alias may be introduced with or without
    the `AS` keyword.

* **WITH ( option_key = option_value [ , ... ] )**

    Specifies dynamic table options for this `UPDATE` operation. These options are passed to the

>     data source connector when writing to the table. The supported options depend on the connector.

Reviewer: Both WITH (option_key = option_value) entries (UPDATE and DELETE FROM) describe the clause generically but don't mention: (a) that these options apply to this statement only, without changing the table's persistent configuration, (b) that a key that isn't a valid identifier needs backtick-quoting, and (c) that options the connector doesn't recognize are silently ignored. All three are worth stating explicitly so users know what to expect.

```suggestion
* **WITH ( key = val [ , ... ] )**

    An optional list of dynamic table options passed to the Data Source V2 connector for this
    statement only. The options are surfaced to the connector's row-level write, allowing
    per-statement tuning (for example a write file size or an isolation level) without changing
    the table configuration. Keys and values are treated as strings; a key that is not a valid
    identifier can be quoted with backticks. Options that the connector does not recognize are
    ignored.
```

* **SET column = expression [ , ... ]**

    Assigns a value to one or more columns. Each value may be an expression or `DEFAULT`. A nested

>     field may be targeted by using a qualified column name.

Reviewer: The SET column = expression entry says a value may be an expression or DEFAULT, which is accurate, but doesn't call out that the expression can be an uncorrelated subquery over a completely different table (e.g. SET salary = (SELECT max(salary) FROM other_table)). Worth adding a sentence here. 

```suggestion
* **SET column = value [ , ... ]**

    Specifies the columns to update and the values to assign to them. Each `value` is an
    expression, typically referencing columns of the target table, but it may also be an
    uncorrelated subquery over other tables. A comma separates each assignment.
```

* **WHERE boolean_expression**

    Specifies an optional condition that selects the rows to update. If the `WHERE` clause is
    omitted, all rows are updated.

### Examples

The following examples assume that an `employees` table has already been created and populated.

#### Update Rows Matching a Condition

```sql
UPDATE employees
    SET salary = salary + 1000
    WHERE department = 'Engineering';
```

#### Update Multiple Columns Using an Alias

```sql
UPDATE employees AS e
    SET e.salary = e.salary * 1.05, e.status = 'reviewed'
    WHERE e.department = 'Sales';
```

#### Update Using Dynamic Table Options

```sql
-- Option names and values are specific to the table's data source connector.
UPDATE employees WITH (`write.split-size` = 10)
    SET status = 'inactive'
    WHERE last_active_date < DATE '2025-01-01';
```

#### Update All Rows

```sql
UPDATE employees SET status = 'active';
```

### Related Statements

* [DELETE FROM statement](sql-ref-syntax-dml-delete-from.html)
* [MERGE INTO statement](sql-ref-syntax-dml-merge-into.html)

> * [SELECT statement](sql-ref-syntax-qry-select.html)

Reviewer: For consistency, link to INSERT statement as well?

---
layout: global
title: UPDATE
displayTitle: UPDATE
license: |
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
---

### Description

The `UPDATE` statement changes the values of columns in rows that satisfy an optional condition.
When no condition is specified, every row is updated.

> `UPDATE` is supported on tables backed by
> [Data Source V2](sql-v2-data-sources.html#row-level-dml) connectors that support row-level
> operations.

Reviewer 2: **Finding 4.** `RewriteUpdateTable.scala:43` calls `checkNoGeneratedColumns(r, UPDATE)`, which fails the statement with `UNSUPPORTED_FEATURE.TABLE_OPERATION` ("... does not support UPDATE TABLE with generated columns") when the table has a generated column *and* declares `GENERATE_COLUMN_VALUES_ON_WRITE`, i.e. asks Spark to compute the value. The comment above `RewriteRowLevelCommand.scala:61` says it's a deliberate fail-fast until recomputation is implemented. Since this sentence is where the page says when `UPDATE` works, one clause here saves a user the error. `MERGE INTO`'s page has the same gap, so a follow-up covering both is fine if you'd rather keep them consistent.

```suggestion
`UPDATE` is supported on tables backed by
[Data Source V2](sql-v2-data-sources.html#row-level-dml) connectors that support row-level
operations. Tables with generated columns whose values Spark computes are not supported, because
Spark cannot yet recompute those values for rewritten rows.
```

### Syntax

```sql
UPDATE table_identifier [ [ AS ] table_alias ]
    [ WITH ( key = value [ , ... ] ) ]
    SET column = value [ , ... ]
    [ WHERE boolean_expression ]
```

### Parameters

* **table_identifier**

>     Specifies the table to update, which may be optionally qualified with a database name.
> 
>     **Syntax:** `[ database_name. ] table_name`

Reviewer 2: **Finding 1.** `UPDATE` only works on Data Source V2 tables, and a V2 table normally lives in a catalog registered through `spark.sql.catalog.*` rather than in the session catalog -- so the form users actually type is `catalog.db.table`, which this syntax doesn't cover. The grammar allows it: `identifierReference` resolves to `multipartIdentifier` (`SqlBaseParser.g4:751`), so any number of name parts is accepted. Several reference pages already use the fuller form, e.g. `docs/sql-ref-syntax-ddl-drop-view.md:44` and `docs/sql-ref-syntax-aux-describe-function.md:52`.

```suggestion
    Specifies the table to update, which may be optionally qualified with a catalog and a database
    name.

    **Syntax:** `[ catalog_name. ] [ database_name. ] table_name`
```

Same on `docs/sql-ref-syntax-dml-delete-from.md:45`.

* **table_alias**

    Specifies an optional alias for the target table. The alias may be introduced with or without
    the `AS` keyword.

* **WITH ( key = value [ , ... ] )**

    Specifies an optional list of dynamic table options passed to the Data Source V2 connector for
    this statement only. The options allow per-statement tuning without changing the table's
    persistent configuration. Keys and values are treated as strings; a key that is not a valid
    identifier can be quoted with backticks. Spark passes options through without validating their

>     names, and connectors may ignore options they do not recognize.

Reviewer 3: Two minor things here.

First, consistency: #57724 documents this same `WITH (...)` clause on the `INSERT`, `MERGE INTO`, and `SELECT` pages in two sentences ("Specifies dynamic table options for this `INSERT` operation. These options are passed to the data source connector when writing to the table. The supported options depend on the connector."). With both PRs landing around the same time, readers will hit the same clause described at two different levels of detail depending on which page they land on. Worth either syncing the wording across all five pages, or describing it once -- the Row-Level DML section of `sql-v2-data-sources.md` that this page already links to is a plausible home -- and giving each page the one-line version plus a link.

Second, if the long form stays, it can lose some weight. The second sentence restates "for this statement only" from the first, and the last sentence makes the same point twice ("passes options through without validating their names" / "connectors may ignore options they do not recognize"). The backtick note is the part that really earns its place, since `write.split-size` in the example below is a parse error unquoted (`propertyKey` is `identifier (DOT identifier)* | stringLit`). One thing that might be worth having in place of "keys and values are treated as strings": keys are case-insensitive, since the options end up in a `CaseInsensitiveStringMap` (`AstBuilder.resolveOptions`). That is behavior a user cannot guess.

* **SET column = value [ , ... ]**

>     Specifies the columns to update and the values to assign to them. Each `value` is an expression,
>     typically referencing columns of the target table, but it may also be `DEFAULT` or an
>     uncorrelated scalar subquery over another table. A comma separates each assignment. A nested
>     field may be targeted by using a qualified column name.

Reviewer 2: **Finding 3.** Each claim here is individually true, but two of them carry a restriction that changes what a reader can write.

`DEFAULT` is only resolved when it is the *whole* assignment value and the key is a top-level column: `ResolveReferencesInUpdate.scala:49-60` passes `defaultReferencesNotAllowedInComplexExpressionsInUpdateSetClause` as the error for anything else, so `SET salary = DEFAULT + 1` is rejected.

"A nested field may be targeted by using a qualified column name" reads as the alias-qualified `e.salary` in the example right below, which is a top-level column, not a nested field. And nested targeting only goes through structs -- `AssignmentUtils.scala:231` errors with "Updating nested fields is only supported for StructType" for a field inside an array or a map.

Two more rules belong in this bullet, both enforced in `AssignmentUtils.scala:175` and `:180`: a column may be assigned at most once, and a column and one of its nested fields cannot both be assigned. And since `RewriteUpdateTable.scala:75` computes every assignment value in one projection over the pre-update row, `SET a = b, b = a` swaps the two columns rather than setting both to the old `b` -- exactly the kind of thing a reference page should pin down.

```suggestion
    Specifies the columns to update and the values to assign to them. Each `value` is an expression,
    typically referencing columns of the target table, but it may also be `DEFAULT` or an
    uncorrelated scalar subquery over another table. `DEFAULT` must be the whole value; it cannot
    appear inside a larger expression. A comma separates each assignment. Every value is evaluated
    against the row as it was before the update, so `SET a = b, b = a` swaps the two columns. A
    column may be assigned at most once, and a column and one of its nested fields cannot both be
    assigned in the same statement. A field of a struct column may be targeted with a dotted path,
    for example `SET address.city = 'Berlin'`; fields nested inside an array or a map cannot be.
```

* **WHERE boolean_expression**

    Specifies an optional condition that selects the rows to update. If the `WHERE` clause is
    omitted, all rows are updated.

### Examples

> The following examples assume that an `employees` table has already been created and populated.

Reviewer 3: The neighbouring DML pages make their examples runnable end to end. `MERGE INTO` opens its Examples section with the initial `SELECT * FROM target` / `SELECT * FROM source` state and shows the resulting table under each example; `INSERT TABLE` creates `students` inline and shows the output after each insert.

Here there is no schema for `employees` and no results, so a reader cannot tell what `status`, `department`, `last_active_date`, or `salary` are, or check that they got the expected outcome. Could you add a small `CREATE TABLE` plus initial `SELECT *` block here, and a result under each example?

#### Update Rows Matching a Condition

```sql
UPDATE employees
    SET salary = salary + 1000
    WHERE department = 'Engineering';
```

#### Update Multiple Columns Using an Alias

```sql
UPDATE employees AS e
    SET e.salary = e.salary * 1.05, e.status = 'reviewed'

>     WHERE e.department = 'Sales';

Reviewer 3: Minor, and it mostly goes away once the example table has a stated schema: if `salary` is an `INT`, then `salary * 1.05` is a `DOUBLE`, and ANSI store assignment permits numeric narrowing (`Cast.canANSIStoreAssign` allows any `NumericType` -> `NumericType`), so the 5% raise is silently truncated back to an `INT`. Giving `salary` a `DECIMAL` or `DOUBLE` type in the setup, or using an increment that stays integral, avoids demonstrating truncation by accident.
